File: design/spikes/SPIKE-004-llm-based-extraction/scripts/spike.mjs
// Throwaway spike script — SPIKE-004 (RFC-003 Appendix §9.4).
// Tests schema-constrained LLM extraction (via @openrouter/sdk) against SPIKE-001's
// stratified sample, on two models, and tests per-clue vs per-puzzle batching.
import { OpenRouter } from "@openrouter/sdk";

const openRouter = new OpenRouter();

const MODELS = {
  frontier: "anthropic/claude-sonnet-4.5",
  cheap: "google/gemini-2.5-flash-lite",
};

async function extract(model, text, schemaName, schema) {
  const result = await openRouter.chat.send({
    chatRequest: {
      model,
      messages: [
        {
          role: "system",
          content: "Extract structured data from the puzzle clue text exactly per the provided JSON schema. Do not infer facts not stated in the text.",
        },
        { role: "user", content: text },
      ],
      responseFormat: {
        type: "json_schema",
        jsonSchema: { name: schemaName, schema, strict: true },
      },
    },
  });
  const content = result.choices[0].message.content;
  try {
    return JSON.parse(content);
  } catch {
    return { _raw: content };
  }
}

async function section(title, model, text, schemaName, schema) {
  console.log(`\n=== ${title} [${model}] ===`);
  console.log(`- "${text.slice(0, 100)}${text.length > 100 ? "..." : ""}"`);
  try {
    const out = await extract(model, text, schemaName, schema);
    console.log("  ->", JSON.stringify(out));
  } catch (err) {
    console.log("  !! error:", err.message);
  }
}

const houseAssignmentSchema = {
  type: "object",
  properties: {
    nationality: { type: "string" },
    house_color: { type: "string" },
  },
  required: ["nationality", "house_color"],
  additionalProperties: false,
};

const relationSchema = {
  type: "object",
  properties: {
    country_a: { type: "string" },
    country_b: { type: "string" },
    relation: { type: "string" },
  },
  required: ["country_a", "country_b", "relation"],
  additionalProperties: false,
};

const metaRuleSchema = {
  type: "object",
  properties: {
    subject: { type: "string" },
    condition: { type: "string" },
    requirement: { type: "string" },
  },
  required: ["subject", "condition", "requirement"],
  additionalProperties: false,
};

const gridPuzzleSchema = {
  type: "object",
  properties: {
    grid_rows: { type: "integer" },
    grid_cols: { type: "integer" },
    target_sum: { type: "integer" },
    given_cells: {
      type: "array",
      items: {
        type: "object",
        properties: { position: { type: "string" }, value: { type: "integer" } },
        required: ["position", "value"],
        additionalProperties: false,
      },
    },
  },
  required: ["grid_rows", "grid_cols", "target_sum", "given_cells"],
  additionalProperties: false,
};

const thresholdRuleSchema = {
  type: "object",
  properties: {
    derived_variable: { type: "string" },
    comparison: { type: "string" },
    outcome: { type: "string" },
  },
  required: ["derived_variable", "comparison", "outcome"],
  additionalProperties: false,
};

const ruleReferenceSchema = {
  type: "object",
  properties: {
    referenced_rules: { type: "array", items: { type: "string" } },
    outcome: { type: "string" },
  },
  required: ["referenced_rules", "outcome"],
  additionalProperties: false,
};

const personFactsSchema = {
  type: "object",
  properties: {
    facts: {
      type: "array",
      items: {
        type: "object",
        properties: { person: { type: "string" }, value: { type: "string" } },
        required: ["person", "value"],
        additionalProperties: false,
      },
    },
  },
  required: ["facts"],
  additionalProperties: false,
};

const dietaryRequirementsSchema = {
  type: "object",
  properties: {
    requirements: {
      type: "array",
      items: {
        type: "object",
        properties: { person: { type: "string" }, requirement: { type: "string" } },
        required: ["person", "requirement"],
        additionalProperties: false,
      },
    },
  },
  required: ["requirements"],
  additionalProperties: false,
};

const restaurantRowSchema = {
  type: "object",
  properties: {
    name: { type: "string" },
    vegan_friendly: { type: "string" },
    nut_free: { type: "string" },
    gluten_free: { type: "string" },
    price: { type: "string" },
  },
  required: ["name", "vegan_friendly", "nut_free", "gluten_free", "price"],
  additionalProperties: false,
};

for (const [tier, model] of Object.entries(MODELS)) {
  console.log(`\n\n########## MODEL TIER: ${tier} (${model}) ##########`);

  // Shape A/B/C/D baseline (PZL-0001) — isolated sentence.
  await section(
    "Shapes A/B/C/D: attribute-assignment (isolated)",
    model,
    "The Englishman lives in the red house.",
    "house_assignment",
    houseAssignmentSchema
  );

  // Shape E: relation + meta-rule (PZL-0005)
  await section(
    "Shape E: relation extraction",
    model,
    "Avalon and Borealis share a border.",
    "relation",
    relationSchema
  );
  await section(
    "Shape E: meta-rule decomposition",
    model,
    "Two countries that share a border must be colored differently.",
    "meta_rule",
    metaRuleSchema
  );

  // Shape F: implicit grid constraints (PZL-0008) — full combined text (needs the whole passage).
  await section(
    "Shape F: implicit grid constraints (full passage)",
    model,
    "Fill a 3-by-3 grid with each of the digits 1 through 9, using each digit exactly once, " +
      "so that every row, every column, and both diagonals add up to the same total: 15. " +
      "The top-left cell is 4. The top-middle cell is 9. The center cell is 5.",
    "grid_puzzle",
    gridPuzzleSchema
  );

  // Shapes H/I: numeric threshold + rule-chain (PZL-0011) — isolated vs combined (batching test).
  await section(
    "Shapes H/I: threshold rule (ISOLATED single sentence)",
    model,
    "If that score is below 600, the loan is Denied.",
    "threshold_rule",
    thresholdRuleSchema
  );
  await section(
    "Shapes H/I: threshold rule (COMBINED with other clues, batching test)",
    model,
    "The couple's credit tier uses the lower of their two individual credit scores. " +
      "If that score is below 600, the loan is Denied. " +
      "If not denied by rules 1-2, and the requested amount is within policy limits for " +
      "their combined income, the loan is Approved. " +
      "Priya's credit score is 680; Sam's credit score is 750.",
    "threshold_rule",
    thresholdRuleSchema
  );
  await section(
    "Shapes H/I: rule cross-reference (COMBINED)",
    model,
    "If not denied by rules 1-2, and the requested amount is within policy limits for " +
      "their combined income, the loan is Approved.",
    "rule_reference",
    ruleReferenceSchema
  );
  await section(
    "Shapes H/I: compound person facts (COMBINED)",
    model,
    "Priya's credit score is 680; Sam's credit score is 750.",
    "person_facts",
    personFactsSchema
  );

  // Shape K: vocabulary-mapping preferences + raw table row (PZL-0013)
  await section(
    "Shape K: dietary requirements (vocabulary mapping)",
    model,
    "Amara is vegan. Ben has a nut allergy and needs a nut-free kitchen. " +
      "Cora needs gluten-free options.",
    "dietary_requirements",
    dietaryRequirementsSchema
  );
  await section(
    "Shape K: raw markdown table row",
    model,
    "| Thai Palace | No | No | Yes | $$ |",
    "restaurant_row",
    restaurantRowSchema
  );
}

File: design/spikes/SPIKE-003-gliner2-capability/scripts/spike.py
"""Throwaway spike script — SPIKE-003 (RFC-003 Appendix Section 9.3).
Tests GLiNER2's native schema-driven extraction against SPIKE-001's stratified sample.
"""
from gliner2 import GLiNER2

print("Loading fastino/gliner2-base-v1 ...")
model = GLiNER2.from_pretrained("fastino/gliner2-base-v1")


def section(title):
    print(f"\n=== {title} ===")


# --- Shapes A/B/C/D (simple) - PZL-0001 ---
section("Shapes A/B/C/D: attribute-assignment, positional, negation, ordering (PZL-0001)")
entities = model.extract_entities(
    "The Englishman lives in the red house. The Ukrainian drinks tea. "
    "The green house is immediately to the right of the ivory house. "
    "The Norwegian lives next to the blue house. Coffee is drunk in the green house.",
    ["person nationality", "beverage", "house color", "spatial relation"],
)
print(entities)

structured = model.extract_json(
    "The Englishman lives in the red house.",
    {"house_assignment": ["nationality::str::the person's nationality",
                           "house_color::str::the color of the house they live in"]},
)
print(structured)

# --- Shape E: relational fact + generative meta-rule - PZL-0005 ---
section("Shape E: symmetric relation + derived meta-rule (PZL-0005)")
rel = model.extract_relations(
    "Avalon and Borealis share a border. Two countries that share a border must be "
    "colored differently. Avalon is colored Red.",
    ["shares border with"],
)
print(rel)
struct_e = model.extract_json(
    "Two countries that share a border must be colored differently.",
    {"coloring_rule": ["subject::str::what the rule applies to",
                        "condition::str::the triggering condition",
                        "requirement::str::what must hold if the condition is true"]},
)
print(struct_e)

# --- Shape F: implicit grid constraints from a named problem type - PZL-0008 ---
section("Shape F: implicit grid constraints from problem framing (PZL-0008)")
struct_f = model.extract_json(
    "Fill a 3-by-3 grid with each of the digits 1 through 9, using each digit exactly once, "
    "so that every row, every column, and both diagonals add up to the same total: 15. "
    "The top-left cell is 4. The top-middle cell is 9. The center cell is 5.",
    {"grid_puzzle": ["grid_size::str::the dimensions of the grid",
                      "target_sum::str::the total each row/column/diagonal must sum to",
                      "given_cells::list::each stated cell position and its value"]},
)
print(struct_f)

# --- Shapes H/I: numeric threshold + derived variable + rule-chain - PZL-0011 ---
section("Shapes H/I: numeric threshold + derived variable + rule chain (PZL-0011)")
struct_hi = model.extract_json(
    "The couple's credit tier uses the lower of their two individual credit scores. "
    "If that score is below 600, the loan is Denied. "
    "If not denied by rules 1-2, and the requested amount is within policy limits for "
    "their combined income, the loan is Approved. "
    "Priya's credit score is 680; Sam's credit score is 750.",
    {"threshold_rule": ["derived_variable::str::how the checked value is computed",
                         "comparison::str::the threshold comparison applied",
                         "outcome::str::the result if the comparison is true"],
     "rule_reference": ["referenced_rules::list::which earlier rules this rule depends on",
                         "outcome::str::the result of this rule"],
     "person_fact": ["person::str::the named person", "value::str::their stated numeric value"]},
)
print(struct_hi)

# --- Shape K: embedded table + vocabulary-mapping preference - PZL-0013 ---
section("Shape K: preference statement + table row (PZL-0013)")
struct_k = model.extract_json(
    "Amara is vegan. Ben has a nut allergy and needs a nut-free kitchen. "
    "Cora needs gluten-free options.",
    {"dietary_requirement": ["person::str::the named person",
                              "requirement::str::their dietary or allergy requirement"]},
)
print(struct_k)

table_row = model.extract_json(
    "| Thai Palace | No | No | Yes | $$ |",
    {"restaurant_row": ["name::str::restaurant name",
                         "vegan_friendly::str::vegan-friendly column value",
                         "nut_free::str::nut-free kitchen column value",
                         "gluten_free::str::gluten-free options column value",
                         "price::str::price column value"]},
)
print(table_row)
